src/main/scala/mesosphere/marathon/core/health/impl/HealthIndex.scala

```diff
@@ -0,0 +1,99 @@
+package mesosphere.marathon
+package core.health.impl
+
+import mesosphere.marathon.core.health.Health
+import mesosphere.marathon.core.instance.Instance
+import mesosphere.marathon.core.task.Task
+
+import scala.collection.concurrent.TrieMap
+
+trait HealthIndex {
+  def remove(key: Instance.Id): Unit
+
+  def remove(key: Task.Id): Unit
+
+  def +=(kv: (Task.Id, Health)): this.type
+
+  def values: Iterable[Health]
+
+  def filterKeys(p: Task.Id => Boolean): collection.Map[Task.Id, Health]
+
+  def get(key: Instance.Id): Option[Health]
+
+  def getOrElse(key: Task.Id, default: => Health): Health
+
+  def retain(p: Task.Id => Boolean): this.type
+}
+
+/** Old implementation, search by Instance.Id is slow */
+case class HealthSingleTrieMap() extends HealthIndex {
```

Comment:
Why do we keep this? Do we even need the base class?

Reply:
Having an interface is always good. The old implementation is there for tests and for reference, but it could be removed.